### PR TITLE
SEQNG-922A: Unit test for the observation state machine

### PR DESCRIPTION
**Pull request by cquiroz**
There is a report of NIRI observations not working after the last updates
This PR adds unit tests to reproduce the issue and tests for NIFS and GMOS. This was a fix can be done without having regressions

These tests require epics but for speed sake the epics layer has been mocked using ScalaMock

### CODE REVIEW

**Comment by cquiroz** [comment]
Yes, we should capture traces of each instrument in all the possible scenarios to reproduce them as tests

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+    observe.onCarClidChange(359)
+    assert(!observe.applyState().isIdle())
+    // CAR VAL change
+    observe.onCarValChange(CarState.BUSY)
+    assert(!observe.applyState().isIdle())
+
+    // OBSERVE goes IDLE
+    // Observe CAR VAL change
+    observe.onObserveCarValChange(CarState.IDLE)
+    // And we are done and IDLE
```
Comment (cquiroz) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: This is a bit odd, should we wait for apply to go IDLE before going back to idleState

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    // CAR VAL change
+    observe.onCarValChange(CarState.BUSY)
+    assert(!observe.applyState().isIdle())
+    // Another VAL change
+    observe.onApplyValChange(4171)
+    assert(!observe.applyState().isIdle())
+    // Observe CAR VAL change
+    observe.onObserveCarValChange(CarState.PAUSED)
+    // We are now idle
+    assert(observe.applyState().isIdle())
```
Comment (cquiroz) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: @jluhrs Does this seem right? Should the apply go back to idle when paused?

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+    observe.onCarClidChange(4171)
+    assert(observe.applyState().isIdle())
+    // CAR VAL change
+    observe.onCarValChange(CarState.IDLE)
+    assert(observe.applyState().isIdle())
+
+    // RESUME OBSERVE
+    observe.onApplyValChange(4172)
+    // TODO Fix why this is idle here, it should be busy
+    assert(observe.applyState().isIdle())
```
Comment (cquiroz) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: @jluhrs I think this is wrong, we should go back to not idle when resumed, right?

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala (hunk 2 above)
Comment (jluhrs) [reply modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: Yes, that is correct.

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala (hunk 3 above)
Comment (jluhrs) [reply modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: Observe and resume work almost the same. Resume will not succeed if the observation was not paused. The same post is used to trigger the resume, and from then it should follow the same sequence of events.

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala (hunk 1 above)
Comment (jluhrs) [reply modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: Yes, you are right

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala (hunk 3 above)
Comment (cquiroz) [reply modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: Perhaps I need to send another post?

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+      "gm:abort",
+      "GMOS Observe",
+      classOf[CarState],
+      epicsService)
+    // Start idle
+    assert(observe.applyState().isIdle())
+    // Post an observe
+    observe.post()
+
+    // CONFIGURATION
```
Comment (jluhrs) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: Configuration is not triggered nor monitored by the `ObserveSenderImpl`, but by an instance of `ApplySenderImpl`. In the case of GMOS, they use the same apply and CAR records, which means the `ObserveSenderImpl` instance will see all the events, but it must stay idle.

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+    observe.onApplyValChange(4166)
+    assert(!observe.applyState().isIdle())
+    // CAR CLID change
+    observe.onCarClidChange(4166)
+    assert(!observe.applyState().isIdle())
+    // CAR VAL change
+    observe.onCarValChange(CarState.IDLE)
+    assert(!observe.applyState().isIdle())
+
+    // OBSERVE
```
Comment (jluhrs) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: The observe must be triggered by a call to `post` in the `ObserveSenderImpl` instance.

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+    observe.onCarValChange(CarState.IDLE)
+    assert(!observe.applyState().isIdle())
+
+    // OBSERVE goes IDLE
+    // Observe CAR VAL change
+    observe.onObserveCarValChange(CarState.IDLE)
+    // And we are done and IDLE
+    assert(observe.applyState().isIdle())
+
+    // ENDOBSERVE
```
Comment (jluhrs) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: endObserve is triggered through the `ApplySenderImpl` instance. This behavior is correct.

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+      "gm:abort",
+      "GMOS Observe",
+      classOf[CarState],
+      epicsService)
+    // Start idle
+    assert(observe.applyState().isIdle())
+    // Post an observe
+    observe.post()
+
+    // CONFIGURATION
```
Comment (jluhrs) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: Again, configuration is triggered through the `ApplySenderImpl` instance, so the `post` in line 230 is wrong. The `ObserveSenderImpl` must remain idle.

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+    assert(observe.applyState().isIdle())
+    // CAR CLID change
+    observe.onCarClidChange(4171)
+    assert(observe.applyState().isIdle())
+    // CAR VAL change
+    observe.onCarValChange(CarState.IDLE)
+    assert(observe.applyState().isIdle())
+
+    // RESUME OBSERVE
+    observe.onApplyValChange(4172)
```
Comment (jluhrs) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: There should be a call to `post` before this.

**Comment by jluhrs** [comment]
You should also include checks for the `CaCommandMonitor` instance returned by post. Specially check that the `CaCommandListener` registered is called once and only once at the end.

File: modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    // CAR CLID change
+    observe.onCarClidChange(4172)
+    assert(!observe.applyState().isIdle())
+    // Apply goes IDLE
+    observe.onCarValChange(CarState.IDLE)
+    assert(!observe.applyState().isIdle())
+
+    // Observe CAR VAL change
+    observe.onObserveCarValChange(CarState.IDLE)
+
```
Comment (jluhrs) [inline modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: `assert(observe.applyState().isIdle())`

Comment (cquiroz) [reply modules/acm/src/test/scala/edu/gemini/epics/acm/ObserveStateSpec.scala]: Updated

**Pull request merged**